src/Raven.Server/ServerWide/ClusterStateMachine.cs

```diff
@@ -1350,23 +1462,105 @@ public IEnumerable<(string Key, long Index)> GetCompareExchangeIndexes(Transacti
             }
         }
 
+        public IEnumerable<(string key, long index, BlittableJsonReaderObject value)> GetCompareExchangeFromPrefix(TransactionOperationContext context, string dbName, long fromIndex, int take)
+        {
+            using (CompareExchangeCommandBase.GetPrefixIndexSlices(context.Allocator, dbName, fromIndex, out var buffer))
+            {
+                var table = context.Transaction.InnerTransaction.OpenTable(CompareExchangeSchema, CompareExchange);
+                using (Slice.External(context.Allocator, buffer, buffer.Length, out var keySlice))
+                using (Slice.External(context.Allocator, buffer, buffer.Length - sizeof(long), out var prefix))
+                {
+                    foreach (var tvr in table.SeekForwardFromPrefix(CompareExchangeSchema.Indexes[CompareExchangeIndexSlice], keySlice, prefix, 0))
+                    {
+                        if (take-- <= 0)
+                            yield break;
+
+                        var key = ReadCompareExchangeKey(tvr.Result.Reader, dbName);
+                        var index = ReadCompareExchangeOrTombstoneIndex(tvr.Result.Reader);
+                        var value = ReadCompareExchangeValue(context, tvr.Result.Reader);
+
+                        yield return (key, index, value);
+                    }
+                }
+            }
+        }
+
+        public IEnumerable<string> GetCompareExchangeTombstonesByKey(TransactionOperationContext context,
+            string dbName, long fromIndex = 0, int take = int.MaxValue)
+        {
+            using (CompareExchangeCommandBase.GetPrefixIndexSlices(context.Allocator, dbName, fromIndex, out var buffer))
+            {
+                var table = context.Transaction.InnerTransaction.OpenTable(CompareExchangeTombstoneSchema, CompareExchangeTombstones);
+                using (Slice.External(context.Allocator, buffer, buffer.Length, out var keySlice))
+                using (Slice.External(context.Allocator, buffer, buffer.Length - sizeof(long), out var prefix))
+                {
+                    foreach (var tvr in table.SeekForwardFromPrefix(CompareExchangeTombstoneSchema.Indexes[CompareExchangeTombstoneIndexSlice], keySlice, prefix, 0))
+                    {
+                        if (take-- <= 0)
+                            yield break;
+
+                        var key = ReadCompareExchangeKey(tvr.Result.Reader, dbName);
+
+                        yield return key;
+                    }
+                }
+            }
+        }
+
+        private void ClearCompareExchangeTombstones(TransactionOperationContext context, string type, BlittableJsonReaderObject cmd, long index, ServerStore serverStore)
+        {
+            try
```

Comment:
Need to do this in batches.
We can do it in the ClusterObserver. The same idea as auto indexes cleaning (every 10 minutes?).

Reply:
done